pkg/containers/containers.go

```diff
@@ -294,6 +295,10 @@ func getContainerIdFromCgroup(cgroupPath string) (string, cruntime.RuntimeId, bo
 			// container root determined by being matched on the last path part
 			return id, runtime, i == len(cgroupParts)-1
 		}
+		if matched := gardenContainerIdFromCgroupRegex.MatchString(id); matched {
+			runtime = cruntime.Garden
```

Comment:
Yep, but there is a problem here, returning a runtime isn't enough for the entire "runtime" logic for tracee. If you look at `pkg/containers/runtime/sockets.go`:

```go
// check default paths for all supported container runtimes and aggregate them
func Autodiscover(onRegisterFail func(err error, runtime RuntimeId, socket string)) Sockets {
	register := func(sockets *Sockets, runtime RuntimeId, socket string) {
		err := sockets.Register(runtime, socket)
		if err != nil {
			onRegisterFail(err, runtime, socket)
		}
	}
	sockets := Sockets{}
	const (
		defaultContainerd = "/var/run/containerd/containerd.sock"
		defaultDocker     = "/var/run/docker.sock"
		defaultCrio       = "/var/run/crio/crio.sock"
		defaultPodman     = "/var/run/podman/podman.sock"
	)

	register(&sockets, Containerd, defaultContainerd)
	register(&sockets, Docker, defaultDocker)
	register(&sockets, Crio, defaultCrio)
	register(&sockets, Podman, defaultPodman)

	return sockets
}
```

we would have to add a socket for the default socket created by garden. (`/var/vcap/data/garden/run/garden.sock` maybe ?). And then we would register this socket.

2nd step would be to create a `ContainerEnricher`. You will find the following ones:

1. pkg/containers/runtime/containerd.go
2. pkg/containers/runtime/crio.go
3. pkg/containers/runtime/docker.go

They all implement the logic to connect to the runtimes and query information about the running containers. This is so tracee can enrich the events with information about the running containers.

If you pick "docker", for example, at `pkg/containers/runtime/docker.go`, it talks to docker socket and queries information about the container given a container ID:

```go
	metadata := ContainerMetadata{
		ContainerId: containerId,
	}
	resp, err := e.client.ContainerInspect(ctx, containerId)
	if err != nil {
		return metadata, errfmt.WrapError(err)
	}
	container := (*resp.ContainerJSONBase)
	metadata.Name = container.Name

	// Docker prefixes a '/' token to local containers.
	// This can cause some confusion so we remove it if relevant.
	if strings.HasPrefix(metadata.Name, "/") {
		metadata.Name = container.Name[1:]
	}

	// get initial image name from docker's container config
	if resp.Config != nil {
		metadata.Image = resp.Config.Image

		// if in k8s extract pod data from the labels
		if resp.Config.Labels != nil {
			labels := resp.Config.Labels
			metadata.Pod = PodMetadata{
				Name:      labels[PodNameLabel],
				Namespace: labels[PodNamespaceLabel],
				UID:       labels[PodUIDLabel],
				Sandbox:   e.isSandbox(labels),
			}
		}
	}
...
```

Reply:
We would have to have a runtime implementation to talk to cloud foundry API (there is a golang API so shouldn't be difficult). Then, through this API we would get information about the container.

Then, like you saw above (for docker), we have to check if cloud foundry sets labels that can be read for us to populate the "k8s" enrichment (Name, Namespace, UID, ...).

This is what will enrich tracee events with information from a specific container! I hope that helps, and, we probably need an escalation for this feature to be done if needed by CNDR (unless someone else implements it, then I can review for sure).

Reply:
This is what will enrich tracee events with information from a specific container! 

Note that, from my quick study about cloud foundry, looks like there might be many combinations to check/support, right ? `garden-k8s`, `garden-runc` (and probably not the `garden-linux` and `guardian` old container managers).

I hope that helps, and, we probably need an escalation for this feature to be done if needed by CNDR (unless someone else implements it, then I can review for sure). Feel free to ping me!

Reply:
@itaysk FYI.

Reply:
Spoke to @hagitaquasec offline and we are gonna flag this as a partial change for garden-runc. This regex change will provide the "container_id" event context only. Any enrichment for garden (or cloud-foundry) would have to be done in another PR/change.

Reply:
> I do know this was under Nadav's radar (specially after https://github.com/aquasecurity/tracee/issues/3003), but I wanted to bring attention to it:

> - I believe our engine list might not be up-to-date with existing container engine setups (and we might need to refactor the code).
> - k8s enrichment should be done talking to k8s API server.

@rafaeldtinoco I've closed the #3003 since it was considering only GKE that I tested and didn't detect any issue. But this is the continuation of the analysis in other env/platforms: https://github.com/aquasecurity/tracee/issues/3003#issuecomment-1771567819

Reply:
Thanks Geyslan, left you a comment there, thanks for all the effort with that.